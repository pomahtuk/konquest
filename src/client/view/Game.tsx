import React, { ReactElement } from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import { startGame } from "../actions/game.actions";
import gameSelectorFunction, { GameStoreSlice } from "../selectors/game.selector";

import Button from "./foundations/Button";
import GameField from "./GameField";

export interface DispatchProps {
  onRestartGame: () => void;
}

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    onRestartGame: (): void => {
      dispatch(startGame());
    }
  };
};

class Game extends React.PureComponent<GameStoreSlice & DispatchProps> {
  render(): ReactElement {
    console.log("re-rendering Game");

    const { onRestartGame } = this.props;

    return (
      <div>
        <GameField />
        <Button variant="destructive" onClick={onRestartGame}>
          Restart game
        </Button>
      </div>
    );
  }
}

export default connect(
  gameSelectorFunction,
  mapDispatchToProps
)(Game);
