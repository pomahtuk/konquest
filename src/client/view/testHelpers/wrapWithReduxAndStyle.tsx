import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
