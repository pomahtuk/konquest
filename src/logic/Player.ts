export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
