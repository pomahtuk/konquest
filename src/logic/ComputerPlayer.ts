import Player, { PlayerTurnOrder } from "./Player";
import { PlanetMap } from "./Planet";
import Fleet from "./Fleet";

export enum ComputerPlayerType {
  EASY = "easy",
  NORMAL = "normal",
  HARD = "hard"
}

class ComputerPlayer extends Player {
  public computerType: ComputerPlayerType;

  constructor(name: string, type: ComputerPlayerType) {
    super(name);
    this.isComputer = true;
    this.computerType = type;
  }

  public shouldSkipPlanet(planetName: string, orders: PlayerTurnOrder[], fleets: Fleet[]): boolean {
    for (const order of orders) {
      if (order.destination === planetName) {
        return true;
      }
    }

    for (const fleet of fleets) {
      if (fleet.destination === planetName) {
        return true;
      }
    }

    return false;
  }

  public takeTurn(planets: PlanetMap, fleets: Fleet[]): PlayerTurnOrder[] {
    return [];
  }
}

export default ComputerPlayer;
