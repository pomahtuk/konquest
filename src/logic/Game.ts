import lzString from "lz-string";

import Planet, { PlanetMap, SerializedPlanetMap } from "./Planet";
import Player, { PlayerTurn, SerializedPlayerTurn } from "./Player";
import Fleet, { SerializedFleet } from "./Fleet";

import getPlanetName from "./helpers/getPlanetName";
import getDistanceBetweenPoints from "./helpers/getDistanceBetweenPoints";
import validateGameParams from "./helpers/validateGameParams";
import validateTurnData from "./helpers/validateTurnData";
import placePlanets from "./helpers/placePlanets";
import conductBattle from "./helpers/conductBattle";
import markDeadPlayers from "./helpers/markDeadPlayers";
import playerMapper from "./helpers/playerMapper";

export interface GameOptions {
  fieldHeight: number;
  fieldWidth: number;
  neutralPlanetCount: number;
  players: Player[];
}

export interface GameData {
  planets: SerializedPlanetMap;
  turns: SerializedPlayerTurn[];
  planetCount: number;
  waitingForPlayer: number;
  fleetTimeline: SerializedFleet[][];
  status: GameStatus;
  winner: Player | null;
}

export enum GameStatus {
  "NOT_STARTED" = "not_started",
  "IN_PROGRESS" = "in_progress",
  "COMPLETED" = "completed"
}

export enum TurnStatus {
  "VALID" = "valid",
  "IGNORED" = "ignored",
  "INVALID" = "invalid"
}

export type GamePersister = (serializedGame: string) => void;

// making fields truly private
const addPlayers = Symbol("addPlayers");
const addNeutralPlanets = Symbol("addNeutralPlanets");
const processTurn = Symbol("processTurn");
const fleetTimeline = Symbol("fleetTimeline");
const waitingForPlayer = Symbol("waitingForPlayer");
const players = Symbol("players");
const planets = Symbol("planets");
const planetCount = Symbol("planetCount");
const fieldHeight = Symbol("fieldHeight");
const fieldWidth = Symbol("fieldWidth");
const turns = Symbol("turns");
const status = Symbol("status");
const winner = Symbol("winner");
const findWinner = Symbol("findWinner");
const findNextValidPlayer = Symbol("findNextValidPlayer");
const addDataToTurn = Symbol("addDataToTurn");

const SymbolMapForSerializer = {
  planets: planets,
  turns: turns,
  planetCount: planetCount,
  waitingForPlayer: waitingForPlayer,
  fleetTimeline: fleetTimeline,
  status: status,
  winner: winner
};

class ConquestGame {
  public static maxSize = 20;
  public static minSize = 4;
  public static minPlayers = 2;
  public static maxPlayers = 4;

  private [fieldHeight]: number;
  private [fieldWidth]: number;
  private [turns]: PlayerTurn[] = [];
  private [planets]: PlanetMap = {};
  private [players]: Player[] = [];
  private [planetCount] = 0;
  private [waitingForPlayer] = 0;
  private [fleetTimeline]: Fleet[][] = [];
  private [status]: GameStatus = GameStatus.NOT_STARTED;
  private [winner]: Player | null = null;
  private persister?: GamePersister;

  public get waitingForPlayer(): number {
    return this[waitingForPlayer];
  }

  public get status(): GameStatus {
    return this[status];
  }

  public get winner(): Player | null {
    return this[winner];
  }

  public get fieldSize(): [number, number] {
    return [this[fieldHeight], this[fieldWidth]];
  }

  public constructor({ fieldHeight: height, fieldWidth: width, neutralPlanetCount, players: newPlayers }: GameOptions, persister?: GamePersister) {
    const result = validateGameParams({ fieldHeight: height, fieldWidth: width, neutralPlanetCount, players: newPlayers });
    if (!result.valid) {
      // TODO: not sure throwing is a best approach here
      throw new Error(result.error);
    }

    this[fieldHeight] = height;
    this[fieldWidth] = width;
    // create game field
    // get players in easy way
    this[addPlayers](newPlayers);
    // add neutral planets
    this[addNeutralPlanets](neutralPlanetCount, newPlayers.length);
    // place planets
    placePlanets({
      planets: this[planets],
      fieldHeight: this[fieldHeight],
      fieldWidth: this[fieldWidth],
      planetCount: this[planetCount]
    });
    this.persister = persister;
    if (this.persister) {
      this.persister(this.serialize());
    }
  }

  public static deSerialize(serializedGame: string, persister?: GamePersister): ConquestGame | undefined {
    const { gameOptions, gameData }: { gameOptions: GameOptions; gameData: GameData } = JSON.parse(
      lzString.decompressFromEncodedURIComponent(serializedGame)
    );

    const playersMap = new Map<string, Player>();
    gameOptions.players = gameOptions.players.map((player) => {
      // TODO: account for computer player
      const newPayer = playerMapper(player, true);
      playersMap.set(player.id, newPayer);
      return newPayer;
    });

    const game = new ConquestGame(gameOptions);

    // and now update game data
    Object.keys(gameData).forEach((dataKey) => {
      let data = gameData[dataKey];
      switch (dataKey) {
        case "turns":
          data = data.map(
            (turn: SerializedPlayerTurn): PlayerTurn => {
              return { orders: turn.orders, player: playersMap.get(turn.playerId) as Player };
            }
          ) as PlayerTurn[];
          break;
        case "planets":
          data = Object.keys(data).reduce((acc, planetKey): PlanetMap => {
            const planet = data[planetKey];
            const newPlanet = new Planet(planetKey, playersMap.get(planet.ownerId), planet.coordinates);
            newPlanet.production = planet.production;
            newPlanet.ships = planet.ships;
            newPlanet.killPercent = planet.killPercent;
            newPlanet.mainColor = planet.mainColor;
            acc[planetKey] = newPlanet;
            return acc;
          }, {});
          break;
        case "fleetTimeline":
          data = data.reduce((acc: Fleet[][], fleets: SerializedFleet[]): Fleet[][] => {
            const newFleets: Fleet[] = fleets.map((fleet: SerializedFleet) => {
              const newFleet = new Fleet({
                owner: playersMap.get(fleet.ownerId) as Player,
                killPercent: fleet.killPercent,
                amount: fleet.amount,
                destination: fleet.destination
              });
              return newFleet;
            });
            acc.push(newFleets);
            return acc;
          }, []);
          break;
        default:
          break;
      }
      game[SymbolMapForSerializer[dataKey]] = data;
    });

    // only add persister once restored
    game.persister = persister;

    return game;
  }

  public addPlayerTurnData(data: PlayerTurn): TurnStatus {
    // do not accept any turns for completed game
    if (this[status] === GameStatus.COMPLETED) {
      return TurnStatus.IGNORED;
    }
    const { player } = data;
    const playerWeAreWaitingFor = this[players][this[waitingForPlayer]];
    if (player.id !== playerWeAreWaitingFor.id) {
      return TurnStatus.INVALID;
    }
    // this will throw if data is invalid
    const result = validateTurnData(data, this[planets]);
    if (!result.valid) {
      return TurnStatus.INVALID;
    }
    // mark game as in progress if it is not started yet
    if (this.status === GameStatus.NOT_STARTED) {
      this[status] = GameStatus.IN_PROGRESS;
    }
    // check if we already have some data for this turn
    this[addDataToTurn](data);
    // update pointer to player we are waiting for
    this[findNextValidPlayer]();
    if (this.persister) {
      this.persister(this.serialize());
    }
    return TurnStatus.VALID;
  }

  public getPlayers(): Player[] {
    // returning deep copy of an object to prevent modification by pointer
    return JSON.parse(JSON.stringify(this[players]));
  }

  public getPlanets(): PlanetMap {
    // returning deep copy of an object to prevent modification by pointer
    return JSON.parse(JSON.stringify(this[planets]));
  }

  public getTurns(): PlayerTurn[] {
    // returning deep copy of an object to prevent modification by pointer
    return JSON.parse(JSON.stringify(this[turns]));
  }

  public getFleets(): Fleet[][] {
    return JSON.parse(JSON.stringify(this[fleetTimeline]));
  }

  public serialize(): string {
    const gameOptions: GameOptions = {
      fieldHeight: this[fieldHeight],
      fieldWidth: this[fieldWidth],
      neutralPlanetCount: this[planetCount] - this[players].length,
      players: this[players]
    };
    const gameData: GameData = {
      planets: Object.keys(this[planets]).reduce((acc, planetKey): SerializedPlanetMap => {
        const planet = this[planets][planetKey];
        const copy = { ...planet, ownerId: (planet.owner && planet.owner.id) || null };
        delete copy.owner;
        acc[planetKey] = copy;
        return acc;
      }, {}), // remove player instance
      turns: this[turns].map(
        (turn): SerializedPlayerTurn => {
          return { orders: turn.orders, playerId: turn.player.id };
        }
      ), // remove  player instance
      fleetTimeline: this[fleetTimeline].reduce((acc: SerializedFleet[][], fleets): SerializedFleet[][] => {
        const newFleets: SerializedFleet[] = fleets.map((fleet) => {
          const copy = { ...fleet, ownerId: fleet.owner.id };
          delete copy.owner;
          return copy;
        });
        acc.push(newFleets);
        return acc;
      }, []), // remove player instance
      planetCount: this[planetCount],
      waitingForPlayer: this[waitingForPlayer],
      status: this[status],
      winner: this[winner]
    };
    return lzString.compressToEncodedURIComponent(JSON.stringify({ gameOptions, gameData }));
  }

  private [addDataToTurn](data: PlayerTurn): void {
    // check if we already have some data for this turn
    const turn = this[turns];
    turn.push(data);
  }

  private [findNextValidPlayer](): void {
    this[waitingForPlayer] += 1;
    if (this[waitingForPlayer] >= this[players].length) {
      // if we made full circle - start anew
      this[waitingForPlayer] = 0;
      // and process current turn
      this[processTurn]();
    }
    const nextPlayer = this[players][this[waitingForPlayer]];
    if (nextPlayer.isDead) {
      this[findNextValidPlayer]();
    }
    // moving computer turn processing here for now, but make sure not to make a turn once game completed
    if (nextPlayer.isComputer && this[status] !== GameStatus.COMPLETED) {
      const orders = nextPlayer.takeTurn(
        this[planets],
        this[fleetTimeline].reduce((acc, fleetList) => {
          acc.concat(fleetList.filter((fleet) => fleet.owner.id === nextPlayer.id));
          return acc;
        }, [])
      );
      this[addDataToTurn]({
        player: nextPlayer,
        orders
      });
      this[findNextValidPlayer]();
    }
  }

  private [processTurn](): void {
    // send fleets
    const currentTurns = this[turns];
    for (const turn of currentTurns) {
      for (const order of turn.orders) {
        const originPlanet = this[planets][order.origin];
        const destinationPlanet = this[planets][order.destination];
        const fleetTravelTime = getDistanceBetweenPoints(originPlanet.coordinates, destinationPlanet.coordinates);
        const fleetTimelineIndex = fleetTravelTime;
        const fleetTimelinePoint = this[fleetTimeline][fleetTimelineIndex] || [];
        fleetTimelinePoint.push(
          new Fleet({
            owner: originPlanet.owner as Player,
            amount: order.amount,
            killPercent: originPlanet.killPercent,
            destination: destinationPlanet.name
          })
        );
        this[fleetTimeline][fleetTimelineIndex] = fleetTimelinePoint;
        originPlanet.ships -= order.amount;
      }
    }

    // check arrival
    const arrivingFleets = this[fleetTimeline].shift() || [];
    for (const fleet of arrivingFleets) {
      const destinationPlanet = this[planets][fleet.destination];
      if (destinationPlanet.owner && destinationPlanet.owner.id === fleet.owner.id) {
        // arriving to own planet, grow fleet
        destinationPlanet.ships += fleet.amount;
      } else {
        conductBattle({
          attackerFleet: fleet,
          defenderPlanet: destinationPlanet
        });
      }
    }

    // do production only for captured planets
    Object.keys(this[planets]).forEach((planetName): void => this[planets][planetName].produce());
    // mark dead players
    markDeadPlayers({
      players: this[players],
      planets: this[planets],
      remainingTimeline: this[fleetTimeline]
    });

    // clean turn data
    this[turns] = [];

    // check if someone won
    this[findWinner]();
  }

  private [findWinner](): void {
    // may be check if all planets are captured?
    const alivePlayers = this[players].filter((player): boolean => !player.isDead);
    // if we have exactly 1 alive player
    if (alivePlayers.length === 1) {
      this[winner] = alivePlayers[0];
      this[status] = GameStatus.COMPLETED;
    }
  }

  private [addPlayers](newPlayers: Player[]): void {
    this[players] = newPlayers;
    // generate player planets
    newPlayers.forEach((player, index): void => {
      const playerPlanet = new Planet(getPlanetName(index), player);
      this[planets][playerPlanet.name] = playerPlanet;
    });
  }

  private [addNeutralPlanets](neutralPlanetCount: number, playersCount: number): void {
    for (let i = 0; i < neutralPlanetCount; i++) {
      const planet = new Planet(getPlanetName(playersCount + i));
      this[planets][planet.name] = planet;
    }
    this[planetCount] = neutralPlanetCount + playersCount;
  }
}

export default ConquestGame;
