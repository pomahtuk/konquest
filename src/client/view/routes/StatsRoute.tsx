import React, { ReactElement, useContext } from "react";
import { RouteComponentProps, Redirect } from "@reach/router";
import { ThemeContext } from "../themes/ThemeProvider";
import { css } from "emotion";
import hexToRgba from "../helpers/hexToRgba";
import { useSelector, shallowEqual } from "react-redux";
import statsRouteSelector from "../../selectors/stats.route.selector";
import { GameStatus } from "../../../logic/Game";
import Player from "../../../logic/Player";
import { SETTINGS } from "../../routeNames";

const StatsRoute: React.SFC<RouteComponentProps> = (): ReactElement => {
  const theme = useContext(ThemeContext);
  const { winner, players, status } = useSelector(statsRouteSelector, shallowEqual);

  if (status !== GameStatus.COMPLETED) {
    return <Redirect to={SETTINGS} noThrow />;
  }

  const statusContainer = css`
    background: rgba(${hexToRgba(theme.colors.background, "0.65")});
    border: 1px solid ${theme.colors.primaryDark};
    padding: ${theme.units.medium} 0;
  `;

  const tableClass = css`
    width: 100%;
    text-align: center;

    tr {
      padding: ${theme.units.small};
    }
  `;
  return (
    <div className={statusContainer}>
      <table className={tableClass}>
        <thead>
          <tr>
            <th>Player</th>
            <th>Ships produced</th>
            <th>Enemy ships destroyed</th>
            <th>Enemy fleets destroyed</th>
          </tr>
        </thead>
        <tbody>
          {players.map(
            (player: Player): ReactElement => (
              <tr key={player.id}>
                <td>
                  {winner ? player.id === winner.id && "ðŸ‘‘ " : player.isDead && "ðŸ’€ "}
                  {player.screenName + (player.isComputer ? " (ðŸ¤–)" : "")}
                </td>
                <td>{player.statShipCount}</td>
                <td>{player.statEnemyShipsDestroyed}</td>
                <td>{player.statEnemyFleetsDestroyed}</td>
              </tr>
            )
          )}
        </tbody>
      </table>
    </div>
  );
};

export default StatsRoute;
