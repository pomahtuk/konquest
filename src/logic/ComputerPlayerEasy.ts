import ComputerPlayer, { ComputerPlayerType } from "./ComputerPlayer";

class ComputerPlayerEasy extends ComputerPlayer {
  protected minimumShips = 20;
  protected shipCountFactor = 2;

  constructor(name: string) {
    super(name, ComputerPlayerType.EASY);
  }
}

export default ComputerPlayerEasy;
