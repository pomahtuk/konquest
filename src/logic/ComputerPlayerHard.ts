import ComputerPlayer, { ComputerPlayerType } from "./ComputerPlayer";

class ComputerPlayerHard extends ComputerPlayer {
  protected minimumShips = 30;
  protected shipCountFactor = 3;

  constructor(name: string) {
    super(name, ComputerPlayerType.HARD);
  }
}

export default ComputerPlayerHard;
