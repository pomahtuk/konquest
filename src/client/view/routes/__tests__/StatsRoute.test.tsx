import React from "react";
import { render } from "@testing-library/react";

import StatsRoute from "../StatsRoute";
import wrapWithReduxAndStyle, { CurrentStore } from "../../testHelpers/wrapWithReduxAndStyle";
import Player from "../../../../logic/Player";
import { GameStatus } from "../../../../logic/Game";

import { Redirect as MockRedirect } from "@reach/router";
import ComputerPlayerEasy from "../../../../logic/ComputerPlayerEasy";
jest.mock("@reach/router", () => {
  const RouterMocks = jest.requireActual("@reach/router");
  return {
    ...RouterMocks,
    Redirect: jest.fn().mockImplementation(() => {
      return <span>Redirect</span>;
    })
  };
});

const testPlayer = new Player("test");

describe("<StatsRoute />", (): void => {
  it("Redirecting to Settings when game is not complete", (): void => {
    const { container } = render(wrapWithReduxAndStyle(<StatsRoute />));
    expect(container).toMatchSnapshot();
    expect(MockRedirect).toHaveBeenCalledTimes(1);
  });

  // Order matters - this should be the last one
  it("Renders without exploding", (): void => {
    CurrentStore.replaceReducer(() => ({
      game: {
        winner: testPlayer,
        activePlayers: [testPlayer],
        status: GameStatus.COMPLETED
      }
    }));
    CurrentStore.dispatch({ type: "some" });
    const { container } = render(wrapWithReduxAndStyle(<StatsRoute />));
    expect(container).toMatchSnapshot();
  });

  it("Renders players in different states (computer, dead without a winner)", (): void => {
    testPlayer.isDead = true;
    const computerPlayer = new ComputerPlayerEasy("test comp");
    CurrentStore.replaceReducer(() => ({
      game: {
        winner: null,
        activePlayers: [testPlayer, computerPlayer],
        status: GameStatus.COMPLETED
      }
    }));
    CurrentStore.dispatch({ type: "some" });
    const { container } = render(wrapWithReduxAndStyle(<StatsRoute />));
    expect(container).toMatchSnapshot();
  });
});
