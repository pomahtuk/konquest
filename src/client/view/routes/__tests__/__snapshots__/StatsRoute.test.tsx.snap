// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatsRoute /> Redirecting to Settings when game is not complete 1`] = `
<div>
  <span>
    Redirect
  </span>
</div>
`;

exports[`<StatsRoute /> Renders players in different states (computer, dead without a winner) 1`] = `
.emotion-1 {
  background: rgba(255,255,255,0.3);
  border-radius: 4px;
  padding: 8px 0;
}

.emotion-0 {
  width: 100%;
  text-align: center;
}

.emotion-0 tr {
  padding: 4px;
}

<div>
  <div
    class="emotion-1"
  >
    <table
      class="emotion-0"
    >
      <thead>
        <tr>
          <th>
            Player
          </th>
          <th>
            Ships produced
          </th>
          <th>
            Enemy ships destroyed
          </th>
          <th>
            Enemy fleets destroyed
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            ðŸ’€ 
            test
          </td>
          <td>
            0
          </td>
          <td>
            0
          </td>
          <td>
            0
          </td>
        </tr>
        <tr>
          <td>
            test comp (ðŸ¤–)
          </td>
          <td>
            0
          </td>
          <td>
            0
          </td>
          <td>
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<StatsRoute /> Renders without exploding 1`] = `
.emotion-1 {
  background: rgba(255,255,255,0.3);
  border-radius: 4px;
  padding: 8px 0;
}

.emotion-0 {
  width: 100%;
  text-align: center;
}

.emotion-0 tr {
  padding: 4px;
}

<div>
  <div
    class="emotion-1"
  >
    <table
      class="emotion-0"
    >
      <thead>
        <tr>
          <th>
            Player
          </th>
          <th>
            Ships produced
          </th>
          <th>
            Enemy ships destroyed
          </th>
          <th>
            Enemy fleets destroyed
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            ðŸ‘‘ 
            test
          </td>
          <td>
            0
          </td>
          <td>
            0
          </td>
          <td>
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
